# FEATURE - user id & net id mapping

blackironj opened the pull request:
- 실제 사용하는 id와 network를 유지하는데 사용하는 id는 같을 필요가
없으며, network를 유지하는데 이용하는 id는 정책에 의하여 변경 될 수
있습니다. (현재 network id는 ip를 이용하도록 함)
- 실제 id는 network routing table을 유지하는데 이용할 필요는 없습니다.
- 따라서 ~~real-id~~  user-id 와 net-id를 mapping 시킬 table이 필요

### 추가사항
  * unordered_map -> id mapping table
  * unmapId, mapId 함수 추가

### 변경사항
  ~~* GeneralService rpc 요청을 받았을때, `context->peer()`에서 ip 주소를
읽어와 id mapping table을 세팅 (현재 network id 는 ip 주소를 사용함)~~
  * GeneralService rpc 요청을 할때 ClientContext에 `net_id`를 metadata에 세팅하여 보내고 ServerContext에서 해당 metadata를 읽어와 id mapping table을 세팅 하도록 함.
  * `KBucket::removeDeadNodes`에서 return 값으로 daed node의 hashed_net_id 리스트를
반환하도록 함
  * `refreshBuckets`시 연결이 끊어진 node에 대해 mapping된 id를 삭제함.
  * `sendMessage` 시 보낼 곳을 찾지 못하면, id_mapping_table에서 삭제

 ### 기타 수정사항
- Network단에서만 사용하는 id type 이름 변경

Commit:
# EDIT - ID Type

- Network단에서만 사용하는 id type 이름 변경

Commit:
# FEATURE - user-id <-> net-id mapping

- 실제 사용하는 id와 network를 유지하는데 사용하는 id는 같을 필요가
없으며, network를 유지하는데 이용하는 id는 정책에 의하여 변경 될 수
있다.
- 실제 id는 network routing table을 유지하는데 이용할 필요는 없다.
- 따라서 user-id 와 net-id를 mapping 시킬 table이 필요

- 추가사항
  * unordered_map -> id mapping table
  * unmapId, mapId 함수 추가

- 변경사항
  * GeneralService rpc 요청을 받았을때, `context->peer()`에서 ip 주소를
읽어와 id mapping table을 세팅 (현재 network id 는 ip 주소를 사용함)
  * `KBucket::removeDeadNodes`에서 return 값으로 daed node의 hashed_net_id 리스트를
반환하도록 함
  * `refreshBuckets`시 mapping된 id를 삭제함.
  * `sendMessage` 시 보낼 곳을 찾지 못하면, id_mapping_table에서 삭제

morishjs commented on src/plugins/net_plugin/kademlia/routing.cpp:
```diff
@@ -210,5 +213,26 @@ std::vector<Node> RoutingTable::findNeighbors(HashedIdType const &id, std::size_
+    m_id_map_mutex.unlock();
+  }
+}
+void RoutingTable::unmapId(const hashed_net_id_type &dead_hashed_ids) {
```
singular 가 맞는 것 같습니다.
~~dead_hashed_ids~~ -> dead_hashed_id

morishjs commented on src/plugins/net_plugin/config/include/network_type.hpp:
```diff
@@ -15,8 +15,9 @@ struct IpEndpoint {
 
-using IdType = std::string;   // 32bytes ID
-using HashedIdType = Hash160; // 32bytes ID를 hash160
+using real_id_type = std::string;
```
real_id 보다는 단순하게 user_id 또는 node_id가 낫지 않을까요? 
직역하면 실제 ID 로 해석될텐데 의도가 불분명한 것 같습니다.

blackironj replied:
user_id가 적절해보이네요. 수정할게요.

morishjs commented on src/plugins/net_plugin/net_plugin.cpp:
```diff
@@ -169,8 +171,13 @@ class NetPluginImpl {
-        bucket.removeDeadNodes();
+        auto dead_node_ids = bucket.removeDeadNodes();
         auto nodes = bucket.selectAliveNodes(false);
+        if (dead_node_ids.has_value()) {
```
175번 코드를 181번코드 위로 옮겨주세요

morishjs commented on src/plugins/net_plugin/rpc_services/rpc_services.cpp:
```diff
@@ -108,13 +111,24 @@ void GeneralService::proceed() {
+            vector<string> splitted_peer;
+            string raw_peer = m_context.peer(); // protocol(ipv4/ipv6) : ip : port
+            boost::split(splitted_peer, raw_peer, boost::is_any_of(":"), boost::token_compress_on);
+            m_routing_table->mapId(real_id, splitted_peer[1]);
```
protocol(ipv4/ipv6) : ip : port 라면

splitted_peer[1] 접근하면 ip만 가져올 것 같은데 맞나요? 

현재 net_id는 ip:port 라서 splitted_peer[1], splitted_peer[2] 가 들어가야 할 것 같은데 말이죠

그리고 단순하게 splitted_peer보다는 splitted
raw_peer 보다는 그냥 address가 나은 것 같아요.

blackironj replied:
위쪽 에서 `net_id = ip` 로 수정했습니다.  그래서 splitted_peer[1] 만 사용하였습니다.
`m_context.peer()` 에서 받아오는  `port` 가 실제 해당 node의 rpc server port를 받는게 아니라 rpc client가 되어서 보낼때 사용한 `random port` 받는것 같습니다. 그래서 위와 같이 수정하였었습니다.
( peer() 함수는 grpc 에서 제공하는 함수입니다) 

그런데, 생각해보니 문제가 있네요. 단순히 `net_id = ip` 로만 했을때,  로컬에서 테스트가 어렵고 공유기 환경에서 또한 문제가 있겠네요.
~~어떻게 해야할지. 좋은 방법이 현재 생각이 나지 않네요.~~
ClientContext의 AddMetaData() 와 ServerContext의 client_metadata()를 활용해 net_id를 세팅하는것도 방법이 될것같습니다.

~~말씀해준대로 변수명 수정하겠습니다.~~

Commit:
# EDIT - get net-id

- ClientContext에서 `net_id` 를 세팅하여 보내고, ServerClient에서
`net_id`를 받도록 함.
- context.peer() -> context.client_metadata()

morishjs commented on src/plugins/net_plugin/kademlia/routing.cpp:
```diff
@@ -210,5 +213,26 @@ std::vector<Node> RoutingTable::findNeighbors(HashedIdType const &id, std::size_
   return neighbors;
 }
 
+void RoutingTable::mapId(const user_id_type &real_id, const net_id_type &net_id) {
```
real_id가 아직도 있네요

morishjs commented on src/plugins/net_plugin/kademlia/routing.cpp:
```diff
@@ -210,5 +213,26 @@ std::vector<Node> RoutingTable::findNeighbors(HashedIdType const &id, std::size_
+  m_id_mapping_table.try_emplace(real_id, Hash<160>::sha1(net_id));
+  m_id_map_mutex.unlock();
+}
+void RoutingTable::unmapId(const user_id_type &real_id) {
```
real_id

morishjs commented on src/plugins/net_plugin/kademlia/routing.cpp:
```diff
@@ -210,5 +213,26 @@ std::vector<Node> RoutingTable::findNeighbors(HashedIdType const &id, std::size_
+  }
+}
+void RoutingTable::unmapId(const hashed_net_id_type &dead_hashed_id) {
+  for (auto &[real_id, net_hased_id] : m_id_mapping_table) {
```
real_id

blackironj replied:
수정하였습니다. 번거롭게해서 죄송합니다.
